회원정보 페이지, 비밀번호 변경 페이지 구현

gayeon4679 opened the pull request:
## 📝 설명
회원정보 페이지, 비밀번호 변경 페이지 UI를 구현하고 API 연결하였습니다.
<!-- PR에 대한 설명입니다. -->

## ✅ PR 유형

- [x] 새로운 기능 추가
- [x] 사용자 UI 디자인 구현

## 💻 작업 내용
- 회원정보 페이지 UI 구현

- 사용자 프로필 정보(이름, 이메일 등) API 연동하여 표시

- 비밀번호 변경 페이지 UI 구현

- 비밀번호 변경 API 연결 및 응답 처리
## 🎨UI
![image](https://github.com/user-attachments/assets/cb9a09ec-b688-4a8b-b3c9-d9dbe89e4876)
![image](https://github.com/user-attachments/assets/eb836fa3-8017-4b4d-965a-5feef3a939aa)

<!-- PR 본문을 입력하세요. -->

## 💬 PR 포인트
- 사용자 프로필 정보(이름, 이메일 등)가 API 응답을 기반으로 올바르게 표시되는지 검토해주세요

- 비밀번호 입력 시 유효성 검사가 올바른지 확인해주세요

- API 요청이 정상적으로 이루어지고, 성공/실패 응답에 따라 적절한 피드백이 toast로 출력되는지 확인해주세요

- trigger를 사용해 몇몇 input을 한 번에 유효성 검사하도록 구현했는데, 혹시 더 좋은 방법이나 개선할 점이 있으면 의견 부탁드립니다

<!-- PR 리뷰 시 공유 사항 또는 유심히 보면 좋을 부분을 설명합니다. -->

gayeon4679 committed:
[🔥remove]: 기존에 만들었던 회원정보수정페이지 제거

gayeon4679 committed:
[🚚rename]:마이페이지를 프로필페이지로 이름 수정

gayeon4679 committed:
[✨feat]: 회원정보수정페이지 UI 구현

gayeon4679 committed:
[🎨design]: 회원정보페이지 UI 구현

기존 회원정보수정페이지에서 수정 가능했던 비밀번호를 따로 분리하여 페이지로 만들었습니다.

gayeon4679 committed:
[✨feat]: 비밀번호 변경 페이지 UI 구현

gayeon4679 committed:
[🐛fix]: 브래드크럼에서 home 클릭 시에 이동하는 url 변경

gayeon4679 committed:
[🐛fix]: 회원가입에서 유효성검사 시에 겹치는 글 제거

gayeon4679 committed:
[♻️refactor]: routes에 passwordChange 추가

gayeon4679 committed:
[🚚rename]: orderlist 파일 위치 변경

gayeon4679 committed:
[🐛fix]: breadcrumb div를 react.fragment로 바꾸어 불필요한 div 수정

gayeon4679 committed:
[✨feat]: passwordChange페이지 api 연결

gayeon4679 committed:
[✨feat]: profile페이지 api 연동

gayeon4679 committed:
[♻️refactor]: formfield 따로 파일 분리

gayeon4679 committed:
[♻️refactor]: 에러,성공 메세지 출력 부분 passwordChange.tsx로 이동

gayeon4679 committed:
[♻️refactor]: 에러 시 홈으로 이동하여 toast로 에러메세지 표시하도록 변경

gayeon4679 committed:
[♻️refactor]: 회원정보 페이지 리팩토링

gayeon4679 committed:
[🐛fix]: trigger를 사용해서 유효성검사 오류 수정

폴더명 PasswordChangePage로 변경,
한 input의 내용이 변경되면 관련된 다른 input에 대한 유효성 검사도 자동으로 수행되도록 수정

gayeon4679 committed:
[🚚rename]: index.ts에서 passwordchange 경로 수정

passwordchange 파일 이름을 passwordChangePage로 변경한 후에 바뀐 이름으로 경로 변경

gayeon4679 committed:
[🐛fix]: api error 메세지 변경

gayeon4679 committed:
[💄style]: main 시맨틱태그 구조 변경

gayeon4679 committed:
[✨feat]: breadcrumbTitle.ts 추가

gayeon4679 committed:
Merge branch 'dev' of https://github.com/High-Coding-org/High-Coding-FE into feat/#51-EditProfiePage

gayeon4679 committed:
[🔥remove]: BreadcrumbTitle.ts 삭제

gayeon4679 committed:
[🐛fix]: import breadcrumbAndTitle 경로 수정

meowTarae commented on src/pages/ProfilePage/ProfilePage.tsx:
```diff
@@ -0,0 +1,48 @@
+      </header>
+      <div className="w-full max-w-[1140px] flex justify-between gap-16 mt-2 p-4">
+        <main className="flex-1">
+          <ProfileList {...data.userInfo} />
```
현재 앱을 실행시켰을 때 크래시가 발생하는 부분은 해당 코드입니다.
data.userInfo에서 userInfo에 대한 값을 읽지 못해 문제가 생긴 것인데요, 때문에 아래와 같은 에러가 발생합니다.
```
Uncaught TypeError: Cannot read properties of undefined (reading 'userInfo')
```

흔히 일어날 수 있는 에러이기에, 이 기회에 한 번 공부해보면 좋을 것 같아요.
- 이 에러는 무엇이며, 왜 나타나는걸까요 ?
- 이러한 현상을 방지하는 여러 방법들을 찾아보세요.

gayeon4679 replied:
이 에러는 3가지 경우에서 나타나는 것 같아요
- 변수에 undefined, null을 할당하는 경우
- 객체에 존재하지 않는 속성에 접근하는 경우
- 비동기 작업이 완료되지 않은 상태에서 데이터에 접근하려고 하는 경우

제가 작성한 코드에서는 UserInfo에 undefined가 할당되서 이런 오류가 나타나네요
지금 코드에서는 if(!data)로 하는게 아니라 if(!data.userInfo)로 해서 userInfo가 undefined인 경우도 처리될 수 있게 해야할 것 같네요 😭

meowTarae replied:
좋은 방법이에요 👍 

혹시 `` ?. `` 에 대해 찾아보시겠어요?
명칭은 Optional Chaining 연산자 입니다.
if 문으로 먼저 한 번 거르더라도, 혹시 모를 에러를 방지하기 위해 요거를 공부해보시고 사용해보시면 좋을 것 같아요.

연산자 찾아보는김에 ||, &&, ?? 에 대해서도 공부해보시면 더 좋을 것 같네요~

gayeon4679 replied:
공식 문서에 "Optional chaining (?.) 연산자는 객체의 속성에 접근하거나 함수를 호출할 때 사용됩니다. 이 연산자를 통해 접근하는 객체나 호출하는 함수가 undefined 또는 null이면, 해당 표현식은 에러를 발생시키지 않고 undefined를 반환하며 실행을 중단합니다."라고 나와 있네요.

건호님 말씀대로, if 문을 따로 사용하는 것보다 ?.을 활용하면 오류를 보다 안전하게 처리할 수 있겠네요.

||, &&, ??에 대해서도 한 번 찾아보도록 하겠습니다!

meowTarae commented on src/pages/ProfilePage/useProfile.ts:
ProfilePage에서 useQuery의 get 요청 중 발생하는 로딩, 결과값, 처리도중의 에러를 변수로 받는 훅 함수를 만들어주셨네요!
별도 훅 함수 파일로 잘 분리를 해주셨는데요, 해당 파일의 위치는 src/hooks/api/useProfile.ts 가 좋을 것 같아요.

gayeon4679 replied:
파일 위치는 수정하겠습니다!

meowTarae commented on src/pages/ProfilePage/useProfile.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const useProfile = () => {
+  const token = localStorage.getItem('authToken');
+  const { isLoading, data, error } = useQuery<IProfile>({
+    queryKey: ['profileData', token],
+    queryFn: async () => await getProfile(token),
+    retry: false,
+  });
+
+  return { isLoading, data, error };
```
해당 코드를 아래와 같이 리팩토링 해볼 수도 있어요.
```
export const useProfile = () => {
  const token = localStorage.getItem('authToken');

  return useQuery({
    queryKey: ['profileData', token],
    queryFn: ()=>getProfile(token),
    retry: false,
  });
};

```

gayeon4679 replied:
이런 식으로 useQuery를 바로 return하는 식으로 리팩토링할 수도 있군요😯

meowTarae commented on src/pages/ProfilePage/useProfile.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const useProfile = () => {
+  const token = localStorage.getItem('authToken');
+  const { isLoading, data, error } = useQuery<IProfile>({
+    queryKey: ['profileData', token],
+    queryFn: async () => await getProfile(token),
+    retry: false,
+  });
+
+  return { isLoading, data, error };
```
useQuery 의 ``error`` 값을 받으셨네요.
현재 error 변수의 역할은
`profilePage에서 데이터를 불러오는 과정에서 에러가 발생하면 홈으로 리다이렉트 시킨다.`
라는 기능을 위해서만 존재하는 것 같은데요.

단순히 에러의 여부만 체크한다면, error 보다는 isError가 어떨까요 ?
그렇다면 error 와 isError 의 차이는 무엇일까요 ? 

공식문서에서 찾아보고 어떤 차이가 있는지 알려주세요~

gayeon4679 replied:
isError는 boolean으로 에러의 발생 여부만을 알려주고 
error는  null | TError로 타입이 정의되어 에러 메세지를 봔환해주네요. 
지금은 에러 메세지를 사용하지 않기 때문에 isError만으로도 충분하네요...

isError로 수정하겠습니다!!

meowTarae replied:
좋습니다 😄

meowTarae commented on src/pages/ProfilePage/useProfile.ts:
```diff
@@ -0,0 +1,15 @@
+import { IProfile } from './type';
+
+export const useProfile = () => {
+  const token = localStorage.getItem('authToken');
+  const { isLoading, data, error } = useQuery<IProfile>({
+    queryKey: ['profileData', token],
```
'authToken', 'profileData' 와 같은 값은 별도 상수 값으로 분리하는게 좋아보여요.

'authToken'의 경우, 제가 '@/constants/localStorageKey' 경로에 LOCAL_STORAGE_AUTH_TOKEN 라는 상수로 선언해놓은게 있으니, 이를 활용해 주세요.

gayeon4679 replied:
넵 수정하겠습니다 !
'profileData'를 상수 값으로 분리하는 것이 좋은 이유가 있을까요?
여러 곳에서 사용하는게 아닌데 상수 값으로 분리해야 한다는게 잘 모르겠어서 물어봅니다..🤔

meowTarae replied:
아무래도 이런 값들은 하드코딩된 상태보다는 상수화하는게 보기가 좋아서 그런 것 같아요.
무조건 빼야한다! 이런건 아니긴 합니다.

meowTarae commented on src/pages/ProfilePage/useProfile.ts:
```diff
@@ -0,0 +1,15 @@
+  const token = localStorage.getItem('authToken');
+  const { isLoading, data, error } = useQuery<IProfile>({
+    queryKey: ['profileData', token],
+    queryFn: async () => await getProfile(token),
```
getProfile 라는 파일은 이미 async 함수인데, 왜 해당 코드에 async await 구문을 적용하셨을까요 ?

gayeon4679 replied:
이건 실수네요 필요없는 코드니까 삭제하도록 하겠습니다!

meowTarae commented on src/services/profile/getProfile.ts:
```diff
@@ -0,0 +1,11 @@
+import axios from 'axios';
+
+export const getProfile = async (token: string) => {
+  const response = await axios.get('https://leoan.p-e.kr/user/info', {
+    headers: {
+      Authorization: `Bearer ${token}`,
+    },
+  });
+
+  return response.data;
+};
```
위 코드도 마찬가지.
response.data라는 코드에 의해 잘못하면 페이지 크래시가 발생할 수 있어요.
이 경우, 코드 전체에 try-catch 문을 적용시켜보는 것은 어떨까요 ?

gayeon4679 replied:
아 그럴 수도 있겠네요... 적용시키도록 하겠습니다.

meowTarae commented on src/services/profile/getProfile.ts:
```diff
@@ -0,0 +1,11 @@
+import axios from 'axios';
+
+export const getProfile = async (token: string) => {
+  const response = await axios.get('https://leoan.p-e.kr/user/info', {
```
<h2>지금처럼 서버 주소를 그대로 드러내면  절대절대절대절대 안됩니다!!!!!!!</h2>

배포 시 악성 해커에 의해 저희 서버가 쉽게 공격 당할 수 있어요.

때문에 저희가  ``.env`` 라는 파일에 서버 주소를 꽁꽁 숨겨서 사용하고 있는거에요.

gayeon4679 replied:
이거 전에 선배가 강조해서 얘기했던 말인데 죄송합니다... 수정하겠습니다

meowTarae commented on src/services/profile/getProfile.ts:
```diff
@@ -0,0 +1,11 @@
+  });
+
+  return response.data;
+};
```
이런 api 요청 코드는 처음에는 작성하기 어려울 수 있으므로, 예제 코드를 남겨놓을게요.

꼭 직접 코드 읽어보시고, 잘 이해가 안되시는 부분 있으면 제게 여쭤봐주세요.

가연님이 작성해주신 코드에서, 개선해볼 수 있는 사항은 다음과 같아요.
1. 제가 만들어놓은 axiosInstance를 사용하여, 서버 url 을 가리고 기타 headers 속성들을 포함한 인스턴스를 사용한다.
2. axiosInstance를 사용하면 apiEndpoint 만 입력하면 되므로 코드가 더 간결해질 수 있다.
3. try-catch 문을 사용하여 api get 요청에 대한 에러만을 위한 에러처리 구문을 만들어보자.
4. api 통신을 통해 전달받은 response 의 타입을 지정하여 더욱 안정적인 데이터 값을 받아볼 수 있다.

```
import { AxiosResponse } from 'axios';

import { API_AUTHORITY, API_ENDPOINT } from '../apiEndpoint';
import { axiosInstance } from '../axiosInstance';

interface IUserInfo {
  birth: string;
  email: string | null;
  id: number;
  name: string;
  password: string | null;
  phoneNumber: string | null;
  username: string;
}

interface IProfileData {
  error: boolean;
  message: string;
  statusCode: number;
  userInfo: IUserInfo;
}

type ProfileResponse = AxiosResponse<IProfileData>;

export const getProfile = async (token: string) => {
  try {
    const res: ProfileResponse = await axiosInstance.get(
      `${API_AUTHORITY.USER}${API_ENDPOINT.PROFILE}`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );

    return res?.data;
  } catch (error) {
    throw new Error(`Error : ${error}`);
  }
};

```

meowTarae reviewed (commented):
가연님! 이제 UI 뿐 아니라 백엔드와의 협업 단계인 API 연결 부분까지 작업해주셨네요.
프론트엔드 개발이 아직 많이 낯설고, 익숙하지 않고, 많이 어려우셨을텐데도 여기까지 작업해주시느라 고생 많으셨어요 😊
또한 pr까지 요청드린 방식으로 잘 수행해주셔서 감사 드립니다.

그런데 하나 말씀드려야 할게 있어요.
몇 일 전에 말씀드렸던 Breadcrumb에 의한 페이지 크래시 된 부분과 
현재 profile 페이지가 크래시되는 부분이 있는데요.
``pr``은 업무처리가 완료된 코드를 머지하기 이전, 마지막으로 팀원들에게 최종 점검을 받는 곳이에요.
그런데 코드를 열람하기도 전에 에러가 발생한다면, 점검을 하기 어렵겠죠?
이점 꼭 유의하셔서 다음번엔 이런 실수는 하지 않기를 바래요.

<hr/>
현재 코드를 확인해본 결과, 수정해야 할 부분들이 많이 보입니다.
회원정보 페이지까지 리뷰를 드릴테니, 이를 참고하여 비밀번호 변경 페이지도 수정 및 코드 점검을 부탁 드리는 바입니다.

meowTarae commented on src/pages/ProfilePage/ProfilePage.tsx:
ProfilePage -> passwordPage 로 이동할 수 있는 방법이 있을까요?
직접적인 url 조작 외에 이동 방법이 필요할 것 같습니다.

gayeon4679 replied:
우측에 있는 사이드바를 통해서 하면 될 것 같습니다.

gayeon4679 committed:
[🚚rename]: useProfile.ts를 src/hooks/api/useProfile.ts 경로로 이동

gayeon4679 committed:
[🐛fix]: data.userInfo에 undefined가 할당되어서 생기는 오류 해결

gayeon4679 committed:
[♻️refactor]: useQuery 바로 return 하도록 수정

gayeon4679 committed:
[♻️refactor]: error를 isError로 변경

gayeon4679 committed:
[♻️refactor]: 'authToken'을 LOCAL_STORAGE_AUTH_TOKEN 상수로 변경

gayeon4679 committed:
[♻️refactor]: getProfile 함수 리팩토링

axios.get을 axiosInstance.get으로 변경하여 API 요청 처리 개선,
하드코딩된 API URL을 API_AUTHORITY와 API_ENDPOINT 상수를 사용한 동적 URL로 변경.

gayeon4679 committed:
[✨feat]: 회원정보 api endpoint 추가

gayeon4679 committed:
[♻️refactor]: profilePage의 type.ts 타입 변경

gayeon4679 committed:
[♻️refactor]: IProfile 타입을 IProfileData로 변경

gayeon4679 committed:
[♻️refactor]: usePasswordChange 훅에서 useMutation을 바로 return하도록 수정

gayeon4679 committed:
[♻️refactor]: changePassword.ts 리팩토링

하드코딩된 서버 url 수정, axios.put을 axiosInstance로 수정, try catch문 추가

gayeon4679 committed:
[✨feat]: 비밀번호변경 api endpoint 추가

gayeon4679 committed:
[🚚rename]: usePasswordChange.ts를 hook으로 이동

gayeon4679 committed:
[♻️refactor]: 하드코딩된 profileDataKey 상수로 변경

gayeon4679 committed:
[♻️refactor]: IPasswordChangeResponse, PasswordChangeData interface 분리, 이름 변경

gayeon4679 committed:
[🐛fix]: 유효성검사 오류 수정

gayeon4679 committed:
[🚚rename]: services/password 에서 services/passwordChange로 폴더이름 변경

gayeon4679 committed:
[♻️refactor]: 비밀번호 변경 리팩토링

token을 changePassword에서 가져와서 사용하는 것으로 수정, onSuccess와 onError로 오류 표시, toast 메세지 수정

gayeon4679 committed:
[🔥remove]: passwordChangedata 타입 제거

gayeon4679 committed:
[✨feat]: 브래드크럼 path에 PROFILE_PASSWORD 추가

gayeon4679 committed:
[♻️refactor]: getProfile.ts에서 token 변수로 선언하여 사용

gayeon4679 committed:
[🐛fix]: 유효성검사에서 trigger 없애고 onSubmit으로 변경

gayeon4679 committed:
[🐛fix]: toast 메시지 출력 데이터 경로 수정

toast 메시지에서 올바른 데이터 경로인 `data.data.message`로 수정

meowTarae committed:
Merge branch 'dev' into feat/#51-EditProfiePage

gayeon4679 committed:
Merge branch 'dev' into feat/#51-EditProfiePage
